ext/ddtrace_profiling_native_extension/extconf.rb

```diff
@@ -192,6 +192,12 @@ def add_compiler_flag(flag)
   $defs << '-DUSE_LEGACY_RB_VM_FRAME_METHOD_ENTRY'
 end
 
+# On older Rubies, rb_gc_force_recycle allowed to free objects in a way that
+# would be invisible to free tracepoints, finalizers and without cleaning
+# obj_to_id_tbl mappings. This leads to potential invisible object re-use
+# to our heap profiler.
+$defs << '-DHAVE_WORKING_RB_GC_FORCE_RECYCLE' if RUBY_VERSION < '3.1'
```

Comment:
So re: yesterday's discussion of how we may be able to workaround the `rb_gc_force_recycle`, I think our intuition was correct (and re https://github.com/ruby/ruby/pull/4363 ).

Here's a testcase (thanks ChatGPT for pairing with me on this):

```diff
--- a/ext/ddtrace_profiling_native_extension/profiling.c
+++ b/ext/ddtrace_profiling_native_extension/profiling.c
@@ -32,6 +32,7 @@ static void holding_the_gvl_signal_handler(DDTRACE_UNUSED int _signal, DDTRACE_U
 static VALUE _native_trigger_holding_the_gvl_signal_handler_on(DDTRACE_UNUSED VALUE _self, VALUE background_thread);
 static VALUE _native_enforce_success(DDTRACE_UNUSED VALUE _self, VALUE syserr_errno, VALUE with_gvl);
 static void *trigger_enforce_success(void *trigger_args);
+static VALUE _native_force_recycle_test(DDTRACE_UNUSED VALUE _self);
 
 void DDTRACE_EXPORT Init_ddtrace_profiling_native_extension(void) {
   VALUE datadog_module = rb_define_module("Datadog");
@@ -65,6 +66,8 @@ void DDTRACE_EXPORT Init_ddtrace_profiling_native_extension(void) {
   rb_define_singleton_method(testing_module, "_native_install_holding_the_gvl_signal_handler", _native_install_holding_the_gvl_signal_handler, 0);
   rb_define_singleton_method(testing_module, "_native_trigger_holding_the_gvl_signal_handler_on", _native_trigger_holding_the_gvl_signal_handler_on, 1);
   rb_define_singleton_method(testing_module, "_native_enforce_success", _native_enforce_success, 2);
+
+  rb_define_singleton_method(testing_module, "_native_force_recycle_test", _native_force_recycle_test, 0);
 }
```

```c
static VALUE _native_force_recycle_test(DDTRACE_UNUSED VALUE _self) {
  int number_of_objs = 6000;
  VALUE ruby_array = rb_ary_new_capa(number_of_objs);
  VALUE object_class = rb_const_get(rb_cObject, rb_intern("Object"));
  VALUE object_ids[number_of_objs];
  VALUE object_ids_after_gc[number_of_objs];

  for (int i = 0; i < number_of_objs; i++) {
    VALUE ruby_object = rb_class_new_instance(0, NULL, object_class);
    rb_ary_store(ruby_array, i, ruby_object);
    object_ids[i] = rb_obj_id(ruby_object);
  }

  for (int i = 0; i < RARRAY_LEN(ruby_array); i++) {
    VALUE ruby_object = rb_ary_entry(ruby_array, i);
    rb_ary_store(ruby_array, i, Qnil);
    rb_gc_force_recycle(ruby_object);
  }

  rb_gc_start();

  for (int i = 0; i < RARRAY_LEN(ruby_array); i++) {
    VALUE new_ruby_object = rb_class_new_instance(0, NULL, object_class);

    if (FL_TEST(new_ruby_object, FL_SEEN_OBJ_ID)) {
      printf("New object at index %d has FL_SEEN_OBJ_ID flag set after creation\n", i);
    }

    rb_ary_store(ruby_array, i, new_ruby_object);

    object_ids_after_gc[i] = rb_obj_id(new_ruby_object);

    if (!FL_TEST(new_ruby_object, FL_SEEN_OBJ_ID)) {
      printf("New object at index %d is missing FL_SEEN_OBJ_ID after rb_obj_id call\n", i);
    }
  }

  for (int i = 0; i < number_of_objs; i++) {
    for (int j = 0; j < number_of_objs; j++) {
      if (object_ids[i] == object_ids_after_gc[j]) {
        printf("Object ID %ld is repeated after garbage collection (Indexes: %d, %d)\n", object_ids[i], i, j);
      }
    }
  }

  return Qtrue;
}
```

and here's the result I get on Ruby 2.7:

```
New object at index 5352 is missing FL_SEEN_OBJ_ID after rb_obj_id call
New object at index 5353 is missing FL_SEEN_OBJ_ID after rb_obj_id call
New object at index 5354 is missing FL_SEEN_OBJ_ID after rb_obj_id call
New object at index 5355 is missing FL_SEEN_OBJ_ID after rb_obj_id call
New object at index 5356 is missing FL_SEEN_OBJ_ID after rb_obj_id call
# ...
New object at index 5995 is missing FL_SEEN_OBJ_ID after rb_obj_id call
New object at index 5996 is missing FL_SEEN_OBJ_ID after rb_obj_id call
New object at index 5997 is missing FL_SEEN_OBJ_ID after rb_obj_id call
New object at index 5998 is missing FL_SEEN_OBJ_ID after rb_obj_id call
New object at index 5999 is missing FL_SEEN_OBJ_ID after rb_obj_id call
Object ID 219601 is repeated after garbage collection (Indexes: 5380, 5999)
Object ID 219641 is repeated after garbage collection (Indexes: 5381, 5998)
Object ID 219681 is repeated after garbage collection (Indexes: 5382, 5997)
Object ID 219721 is repeated after garbage collection (Indexes: 5383, 5996)
Object ID 219761 is repeated after garbage collection (Indexes: 5384, 5995)
# ...
Object ID 244201 is repeated after garbage collection (Indexes: 5995, 5356)
Object ID 244241 is repeated after garbage collection (Indexes: 5996, 5355)
Object ID 244281 is repeated after garbage collection (Indexes: 5997, 5354)
Object ID 244321 is repeated after garbage collection (Indexes: 5998, 5353)
Object ID 244361 is repeated after garbage collection (Indexes: 5999, 5352)
```

...which seems to match with our expectations:

1. Object ids get repeated for objects that get GC'd with `rb_gc_force_recycle`
2. New objects born in their place do not have the `FL_SEEN_OBJ_ID` set
3. Buggy objects never get the `FL_SEEN_OBJ_ID` set after a call to `rb_obj_id`